File: src/main/java/frc/robot/subsystems/climber/ClimberIOReal.java
package frc.robot.subsystems.climber;

import com.revrobotics.CANSparkBase;
import com.revrobotics.CANSparkMax;
import com.revrobotics.RelativeEncoder;
import edu.wpi.first.math.util.Units;
import frc.robot.Constants;

public abstract class ClimberIOReal extends ClimberIO {
    private final CANSparkMax motor = getMotor();
    private final RelativeEncoder encoder = motor.getEncoder();

    protected abstract CANSparkMax getMotor();

    public ClimberIOReal() {
        motor.restoreFactoryDefaults();
        motor.setIdleMode(CANSparkBase.IdleMode.kBrake);
        motor.setCANTimeout(250);
        motor.enableVoltageCompensation(12.0);
        motor.setSmartCurrentLimit(40);
        motor.burnFlash();
        encoder.setPosition(0);
    }

    @Override
    public void updateInputs(ClimberIOInputs inputs) {
        inputs.appliedVolts = motor.getAppliedOutput() * motor.getBusVoltage();
        inputs.currentAmps = motor.getOutputCurrent();
        inputs.positionRad = Units.rotationsToRadians(encoder.getPosition() / Constants.Climber.gearRatio);
    }

    @Override
    public void set(double volts) {
        motor.setVoltage(volts);
    }

    @Override
    public void stop() {
        motor.stopMotor();
    }

    @Override
    public void resetPosition() {
        encoder.setPosition(0);
    }
}

File: src/main/java/frc/robot/subsystems/climber/ClimberIOReal.kt
package frc.robot.subsystems.climber

import com.revrobotics.CANSparkBase
import com.revrobotics.CANSparkMax
import com.revrobotics.RelativeEncoder
import edu.wpi.first.math.util.Units
import frc.robot.Constants

abstract class ClimberIOReal : ClimberIO() {
    private val motor = getMotor()
    private val encoder: RelativeEncoder = motor.encoder

    protected abstract fun getMotor(): CANSparkMax

    init {
        motor.restoreFactoryDefaults()
        motor.setIdleMode(CANSparkBase.IdleMode.kBrake)
        motor.setCANTimeout(250)
        motor.enableVoltageCompensation(12.0)
        motor.setSmartCurrentLimit(40)
        motor.burnFlash()
        encoder.setPosition(0.0)
    }

    override fun updateInputs(inputs: ClimberIOInputs) {
        inputs.appliedVolts = motor.appliedOutput * motor.busVoltage
        inputs.currentAmps = motor.outputCurrent
        inputs.positionRad = Units.rotationsToRadians(encoder.position / Constants.Climber.gearRatio)
    }

    override fun set(volts: Double) {
        motor.setVoltage(volts)
    }

    override fun stop() {
        motor.stopMotor()
    }

    override fun resetPosition() {
        encoder.setPosition(0.0)
    }
}
